azurerm/resource_arm_monitor_diagnostic_setting.go

```diff
@@ -0,0 +1,441 @@
+			Enabled:  utils.Bool(enabled),
+			RetentionPolicy: &insights.RetentionPolicy{
+				Days:    utils.Int32(int32(retentionDays)),
+				Enabled: utils.Bool(retentionEnabled),
+			},
+		}
+
+		results = append(results, output)
+	}
+
+	return results
+}
+
+func flattenMonitorDiagnosticLogs(input *[]insights.LogSettings) []interface{} {
+	results := make([]interface{}, 0)
+	if input == nil {
+		return results
+	}
+
+	for _, v := range *input {
+		output := make(map[string]interface{})
+
+		if v.Category != nil {
+			output["category"] = *v.Category
+		}
+
+		if v.Enabled != nil {
+			output["enabled"] = *v.Enabled
+		}
+
+		policies := make([]interface{}, 0)
+
+		if inputPolicy := v.RetentionPolicy; inputPolicy != nil {
+			outputPolicy := make(map[string]interface{})
+
+			if inputPolicy.Days != nil {
+				outputPolicy["days"] = int(*inputPolicy.Days)
+			}
+
+			if inputPolicy.Enabled != nil {
+				outputPolicy["enabled"] = *inputPolicy.Enabled
+			}
+
+			policies = append(policies, outputPolicy)
+		}
+
+		output["retention_policy"] = policies
+
+		results = append(results, output)
+	}
+
+	return results
+}
+
+func expandMonitorDiagnosticsSettingsMetrics(input []interface{}) []insights.MetricSettings {
+	results := make([]insights.MetricSettings, 0)
+
+	for _, raw := range input {
+		v := raw.(map[string]interface{})
+
+		category := v["category"].(string)
+		enabled := v["enabled"].(bool)
+
+		policiesRaw := v["retention_policy"].([]interface{})
+		policyRaw := policiesRaw[0].(map[string]interface{})
+
+		retentionDays := policyRaw["days"].(int)
+		retentionEnabled := policyRaw["enabled"].(bool)
+
+		output := insights.MetricSettings{
+			Category: utils.String(category),
+			Enabled:  utils.Bool(enabled),
+			RetentionPolicy: &insights.RetentionPolicy{
+				Days:    utils.Int32(int32(retentionDays)),
+				Enabled: utils.Bool(retentionEnabled),
+			},
+		}
+
+		results = append(results, output)
+	}
+
+	return results
+}
+
+func flattenMonitorDiagnosticMetrics(input *[]insights.MetricSettings) []interface{} {
+	results := make([]interface{}, 0)
+	if input == nil {
+		return results
+	}
+
+	for _, v := range *input {
+		output := make(map[string]interface{})
+
+		if v.Category != nil {
+			output["category"] = *v.Category
+		}
+
+		if v.Enabled != nil {
+			output["enabled"] = *v.Enabled
+		}
+
+		policies := make([]interface{}, 0)
+
+		if inputPolicy := v.RetentionPolicy; inputPolicy != nil {
+			outputPolicy := make(map[string]interface{})
+
+			if inputPolicy.Days != nil {
+				outputPolicy["days"] = int(*inputPolicy.Days)
+			}
+
+			if inputPolicy.Enabled != nil {
+				outputPolicy["enabled"] = *inputPolicy.Enabled
+			}
+
+			policies = append(policies, outputPolicy)
+		}
+
+		output["retention_policy"] = policies
+
+		results = append(results, output)
+	}
+
+	return results
+}
+
+type monitorDiagnosticId struct {
+	resourceID string
+	name       string
+}
+
+func parseMonitorDiagnosticId(monitorId string) (*monitorDiagnosticId, error) {
```

Comment:
Might it be better to return two strings rather than a bespoke struck here? This seems the less common method in the provider

Reply:
I think we'd be better to move towards structs in the longer term instead vs returning `(string, string, error)` makes this less clear imo?

Reply:
👍, def should get a better mechanism for parsing IDs into structs sooner then later